File: src/config/presets.ts
/**
 * Preset Configuration
 * 
 * Defines the 11 approved bout formats with their turn orchestration rules.
 * 
 * Turn Types:
 * - alternating: A ↔ B (2-bot ping-pong)
 * - round_robin: A → B → C → A... (rotating sequence)
 * - broadcast: Presenter → [Responders] → Presenter... (asymmetric)
 */

export type TurnType = 'alternating' | 'round_robin' | 'broadcast';

export interface PresetAgent {
  role: string;
  description: string;
  isPresenter?: boolean; // For broadcast: marks the lead role
}

export interface PresetConfig {
  id: string;
  name: string;
  description: string;
  agentCount: number;
  turnType: TurnType;
  agents: PresetAgent[];
  defaultRounds: number;
  freeTier: boolean; // Available on free tier (2-bot presets)
}

export const PRESETS: Record<string, PresetConfig> = {
  // ============================================
  // ALTERNATING (2-bot)
  // ============================================
  
  roast_battle: {
    id: 'roast_battle',
    name: 'Roast Battle',
    description: 'Two comedians trade insults in escalating rounds.',
    agentCount: 2,
    turnType: 'alternating',
    agents: [
      { role: 'roaster_a', description: 'First comedian' },
      { role: 'roaster_b', description: 'Second comedian' },
    ],
    defaultRounds: 4,
    freeTier: true,
  },

  on_the_couch: {
    id: 'on_the_couch',
    name: 'On The Couch',
    description: 'Therapy session between analyst and patient.',
    agentCount: 2,
    turnType: 'alternating',
    agents: [
      { role: 'therapist', description: 'The analyst' },
      { role: 'patient', description: 'The analysand' },
    ],
    defaultRounds: 6,
    freeTier: true,
  },

  gloves_off: {
    id: 'gloves_off',
    name: 'Gloves Off',
    description: 'No-holds-barred debate between opposing viewpoints.',
    agentCount: 2,
    turnType: 'alternating',
    agents: [
      { role: 'debater_a', description: 'First debater' },
      { role: 'debater_b', description: 'Second debater' },
    ],
    defaultRounds: 4,
    freeTier: true,
  },

  // ============================================
  // ROUND ROBIN (3-4 bot, equal participants)
  // ============================================

  darwin_special: {
    id: 'darwin_special',
    name: 'Darwin Special',
    description: 'Darwin observes and discusses specimens with fellow naturalists.',
    agentCount: 3,
    turnType: 'round_robin',
    agents: [
      { role: 'darwin', description: 'Charles Darwin' },
      { role: 'naturalist_1', description: 'First naturalist' },
      { role: 'naturalist_2', description: 'Second naturalist' },
    ],
    defaultRounds: 4,
    freeTier: false,
  },

  first_contact: {
    id: 'first_contact',
    name: 'First Contact',
    description: 'Humanity meets alien intelligence for the first time.',
    agentCount: 3,
    turnType: 'round_robin',
    agents: [
      { role: 'human', description: 'Human representative' },
      { role: 'alien_1', description: 'First alien entity' },
      { role: 'alien_2', description: 'Second alien entity' },
    ],
    defaultRounds: 4,
    freeTier: false,
  },

  writers_room: {
    id: 'writers_room',
    name: 'Writers Room',
    description: 'TV writers brainstorm and pitch ideas.',
    agentCount: 3,
    turnType: 'round_robin',
    agents: [
      { role: 'writer_a', description: 'First writer' },
      { role: 'writer_b', description: 'Second writer' },
      { role: 'writer_c', description: 'Third writer' },
    ],
    defaultRounds: 4,
    freeTier: false,
  },

  the_flatshare: {
    id: 'the_flatshare',
    name: 'The Flatshare',
    description: 'Roommates navigate domestic drama and shared living.',
    agentCount: 3,
    turnType: 'round_robin',
    agents: [
      { role: 'flatmate_a', description: 'First flatmate' },
      { role: 'flatmate_b', description: 'Second flatmate' },
      { role: 'flatmate_c', description: 'Third flatmate' },
    ],
    defaultRounds: 4,
    freeTier: false,
  },

  last_supper: {
    id: 'last_supper',
    name: 'Last Supper',
    description: 'Historical figures share a final meal and conversation.',
    agentCount: 4,
    turnType: 'round_robin',
    agents: [
      { role: 'host', description: 'The host figure' },
      { role: 'guest_1', description: 'First guest' },
      { role: 'guest_2', description: 'Second guest' },
      { role: 'guest_3', description: 'Third guest' },
    ],
    defaultRounds: 4,
    freeTier: false,
  },

  the_mansion: {
    id: 'the_mansion',
    name: 'The Mansion',
    description: 'Murder mystery dinner party with secrets and accusations.',
    agentCount: 4,
    turnType: 'round_robin',
    agents: [
      { role: 'host', description: 'The mansion host' },
      { role: 'guest_1', description: 'First guest' },
      { role: 'guest_2', description: 'Second guest' },
      { role: 'guest_3', description: 'Third guest' },
    ],
    defaultRounds: 4,
    freeTier: false,
  },

  // ============================================
  // BROADCAST (4-bot, asymmetric presenter)
  // ============================================

  shark_pit: {
    id: 'shark_pit',
    name: 'Shark Pit',
    description: 'Entrepreneur pitches to a panel of ruthless investors.',
    agentCount: 4,
    turnType: 'broadcast',
    agents: [
      { role: 'pitcher', description: 'The entrepreneur', isPresenter: true },
      { role: 'shark_1', description: 'First investor' },
      { role: 'shark_2', description: 'Second investor' },
      { role: 'shark_3', description: 'Third investor' },
    ],
    defaultRounds: 3,
    freeTier: false,
  },

  the_summit: {
    id: 'the_summit',
    name: 'The Summit',
    description: 'World leaders debate global policy with high stakes.',
    agentCount: 4,
    turnType: 'broadcast',
    agents: [
      { role: 'proposer', description: 'The proposal author', isPresenter: true },
      { role: 'delegate_1', description: 'First delegate' },
      { role: 'delegate_2', description: 'Second delegate' },
      { role: 'delegate_3', description: 'Third delegate' },
    ],
    defaultRounds: 4,
    freeTier: false,
  },
};

// Utility functions

export function getPreset(id: string): PresetConfig | undefined {
  return PRESETS[id];
}

export function getFreeTierPresets(): PresetConfig[] {
  return Object.values(PRESETS).filter(p => p.freeTier);
}

export function getPremiumPresets(): PresetConfig[] {
  return Object.values(PRESETS).filter(p => !p.freeTier);
}

export function getPresetsByTurnType(turnType: TurnType): PresetConfig[] {
  return Object.values(PRESETS).filter(p => p.turnType === turnType);
}

File: be/src/pit_api/engine/presets.py
"""
Preset Configuration

Defines the 11 approved bout formats with their turn orchestration rules.

Turn Types:
- alternating: A ↔ B (2-bot ping-pong)
- round_robin: A → B → C → A... (rotating sequence)
- broadcast: Presenter → [Responders] → Presenter... (asymmetric)
"""

from dataclasses import dataclass
from enum import Enum
from typing import Optional, TypedDict


class TurnType(str, Enum):
    ALTERNATING = "alternating"
    ROUND_ROBIN = "round_robin"
    BROADCAST = "broadcast"


class PresetAgent(TypedDict):
    role: str
    description: str
    is_presenter: Optional[bool]


@dataclass
class PresetConfig:
    id: str
    name: str
    description: str
    agent_count: int
    turn_type: TurnType
    agents: list[PresetAgent]
    default_rounds: int
    free_tier: bool


# ============================================
# PRESET DEFINITIONS
# ============================================

PRESETS: dict[str, PresetConfig] = {
    # ALTERNATING (2-bot)
    "roast_battle": PresetConfig(
        id="roast_battle",
        name="Roast Battle",
        description="Two comedians trade insults in escalating rounds.",
        agent_count=2,
        turn_type=TurnType.ALTERNATING,
        agents=[
            {"role": "roaster_a", "description": "First comedian", "is_presenter": None},
            {"role": "roaster_b", "description": "Second comedian", "is_presenter": None},
        ],
        default_rounds=4,
        free_tier=True,
    ),

    "on_the_couch": PresetConfig(
        id="on_the_couch",
        name="On The Couch",
        description="Therapy session between analyst and patient.",
        agent_count=2,
        turn_type=TurnType.ALTERNATING,
        agents=[
            {"role": "therapist", "description": "The analyst", "is_presenter": None},
            {"role": "patient", "description": "The analysand", "is_presenter": None},
        ],
        default_rounds=6,
        free_tier=True,
    ),

    "gloves_off": PresetConfig(
        id="gloves_off",
        name="Gloves Off",
        description="No-holds-barred debate between opposing viewpoints.",
        agent_count=2,
        turn_type=TurnType.ALTERNATING,
        agents=[
            {"role": "debater_a", "description": "First debater", "is_presenter": None},
            {"role": "debater_b", "description": "Second debater", "is_presenter": None},
        ],
        default_rounds=4,
        free_tier=True,
    ),

    # ROUND ROBIN (3-4 bot, equal participants)
    "darwin_special": PresetConfig(
        id="darwin_special",
        name="Darwin Special",
        description="Darwin observes and discusses specimens with fellow naturalists.",
        agent_count=3,
        turn_type=TurnType.ROUND_ROBIN,
        agents=[
            {"role": "darwin", "description": "Charles Darwin", "is_presenter": None},
            {"role": "naturalist_1", "description": "First naturalist", "is_presenter": None},
            {"role": "naturalist_2", "description": "Second naturalist", "is_presenter": None},
        ],
        default_rounds=4,
        free_tier=False,
    ),

    "first_contact": PresetConfig(
        id="first_contact",
        name="First Contact",
        description="Humanity meets alien intelligence for the first time.",
        agent_count=3,
        turn_type=TurnType.ROUND_ROBIN,
        agents=[
            {"role": "human", "description": "Human representative", "is_presenter": None},
            {"role": "alien_1", "description": "First alien entity", "is_presenter": None},
            {"role": "alien_2", "description": "Second alien entity", "is_presenter": None},
        ],
        default_rounds=4,
        free_tier=False,
    ),

    "writers_room": PresetConfig(
        id="writers_room",
        name="Writers Room",
        description="TV writers brainstorm and pitch ideas.",
        agent_count=3,
        turn_type=TurnType.ROUND_ROBIN,
        agents=[
            {"role": "writer_a", "description": "First writer", "is_presenter": None},
            {"role": "writer_b", "description": "Second writer", "is_presenter": None},
            {"role": "writer_c", "description": "Third writer", "is_presenter": None},
        ],
        default_rounds=4,
        free_tier=False,
    ),

    "the_flatshare": PresetConfig(
        id="the_flatshare",
        name="The Flatshare",
        description="Roommates navigate domestic drama and shared living.",
        agent_count=3,
        turn_type=TurnType.ROUND_ROBIN,
        agents=[
            {"role": "flatmate_a", "description": "First flatmate", "is_presenter": None},
            {"role": "flatmate_b", "description": "Second flatmate", "is_presenter": None},
            {"role": "flatmate_c", "description": "Third flatmate", "is_presenter": None},
        ],
        default_rounds=4,
        free_tier=False,
    ),

    "last_supper": PresetConfig(
        id="last_supper",
        name="Last Supper",
        description="Historical figures share a final meal and conversation.",
        agent_count=4,
        turn_type=TurnType.ROUND_ROBIN,
        agents=[
            {"role": "host", "description": "The host figure", "is_presenter": None},
            {"role": "guest_1", "description": "First guest", "is_presenter": None},
            {"role": "guest_2", "description": "Second guest", "is_presenter": None},
            {"role": "guest_3", "description": "Third guest", "is_presenter": None},
        ],
        default_rounds=4,
        free_tier=False,
    ),

    "the_mansion": PresetConfig(
        id="the_mansion",
        name="The Mansion",
        description="Murder mystery dinner party with secrets and accusations.",
        agent_count=4,
        turn_type=TurnType.ROUND_ROBIN,
        agents=[
            {"role": "host", "description": "The mansion host", "is_presenter": None},
            {"role": "guest_1", "description": "First guest", "is_presenter": None},
            {"role": "guest_2", "description": "Second guest", "is_presenter": None},
            {"role": "guest_3", "description": "Third guest", "is_presenter": None},
        ],
        default_rounds=4,
        free_tier=False,
    ),

    # BROADCAST (4-bot, asymmetric presenter)
    "shark_pit": PresetConfig(
        id="shark_pit",
        name="Shark Pit",
        description="Entrepreneur pitches to a panel of ruthless investors.",
        agent_count=4,
        turn_type=TurnType.BROADCAST,
        agents=[
            {"role": "pitcher", "description": "The entrepreneur", "is_presenter": True},
            {"role": "shark_1", "description": "First investor", "is_presenter": False},
            {"role": "shark_2", "description": "Second investor", "is_presenter": False},
            {"role": "shark_3", "description": "Third investor", "is_presenter": False},
        ],
        default_rounds=3,
        free_tier=False,
    ),

    "the_summit": PresetConfig(
        id="the_summit",
        name="The Summit",
        description="World leaders debate global policy with high stakes.",
        agent_count=4,
        turn_type=TurnType.BROADCAST,
        agents=[
            {"role": "proposer", "description": "The proposal author", "is_presenter": True},
            {"role": "delegate_1", "description": "First delegate", "is_presenter": False},
            {"role": "delegate_2", "description": "Second delegate", "is_presenter": False},
            {"role": "delegate_3", "description": "Third delegate", "is_presenter": False},
        ],
        default_rounds=4,
        free_tier=False,
    ),
}


# ============================================
# UTILITY FUNCTIONS
# ============================================

def get_preset(preset_id: str) -> PresetConfig | None:
    """Get preset config by ID."""
    return PRESETS.get(preset_id)


def get_free_tier_presets() -> list[PresetConfig]:
    """Get all presets available on free tier."""
    return [p for p in PRESETS.values() if p.free_tier]


def get_premium_presets() -> list[PresetConfig]:
    """Get all premium presets."""
    return [p for p in PRESETS.values() if not p.free_tier]


def get_presets_by_turn_type(turn_type: TurnType) -> list[PresetConfig]:
    """Get all presets with a specific turn type."""
    return [p for p in PRESETS.values() if p.turn_type == turn_type]
